From 680ccef9491c88261b2cef3a785edeba8c2f6d35 Mon Sep 17 00:00:00 2001
From: Matthew Schwartz <matthew.schwartz@linux.dev>
Date: Sat, 7 Dec 2024 11:59:12 -0800
Subject: [PATCH] test disable VK_KHR_present_wait on NV

---
 libs/vkd3d/device.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libs/vkd3d/device.c b/libs/vkd3d/device.c
index 167f126a..183e211b 100644
--- a/libs/vkd3d/device.c
+++ b/libs/vkd3d/device.c
@@ -1531,6 +1531,15 @@ static void vkd3d_physical_device_info_apply_workarounds(struct vkd3d_physical_d
             device->device_info.swapchain_maintenance1_features.swapchainMaintenance1 = VK_FALSE;
             device->vk_info.EXT_swapchain_maintenance1 = false;
         }
+
+        // Workaround for broken VK_KHR_present_wait on NVIDIA under gamescope
+        if (info->vulkan_1_2_properties.driverID == VK_DRIVER_ID_NVIDIA_PROPRIETARY &&
+            info->present_wait_features.presentWait)
+        {
+            WARN("Disabling VK_KHR_present_wait on NVIDIA due to known issues.\n");
+            device->device_info.present_wait_features.presentWait = VK_FALSE;
+            device->vk_info.KHR_present_wait = false;
+        }
     }
 }
 
-- 
2.47.1

